source/common/common/matchers.h

```diff
@@ -191,18 +220,25 @@ class MetadataMatcher {
 
 class PathMatcher : public StringMatcher {
 public:
-  PathMatcher(const envoy::type::matcher::v3::PathMatcher& path) : matcher_(path.path()) {}
-  PathMatcher(const envoy::type::matcher::v3::StringMatcher& matcher) : matcher_(matcher) {}
+  PathMatcher(const envoy::type::matcher::v3::PathMatcher& path)
+      : matcher_(path.path()), pattern_based_match_(false) {}
+  PathMatcher(const envoy::type::matcher::v3::StringMatcher& matcher)
+      : matcher_(matcher), pattern_based_match_(false) {}
+  PathMatcher(const envoy::type::matcher::v3::PatternMatcher& matcher)
+      : pattern_matcher_(matcher), pattern_based_match_(true) {}
 
   static PathMatcherConstSharedPtr createExact(const std::string& exact, bool ignore_case);
   static PathMatcherConstSharedPtr createPrefix(const std::string& prefix, bool ignore_case);
+  static PathMatcherConstSharedPtr createPattern(const std::string& pattern, bool ignore_case);
   static PathMatcherConstSharedPtr
   createSafeRegex(const envoy::type::matcher::v3::RegexMatcher& regex_matcher);
 
   bool match(const absl::string_view path) const override;
 
 private:
   const StringMatcherImpl<envoy::type::matcher::v3::StringMatcher> matcher_;
+  const PatternMatcherImpl<envoy::type::matcher::v3::PatternMatcher> pattern_matcher_;
+  const bool pattern_based_match_;
```

Comment:
nit: should probably be an enum instead of a bool.

Another alternative is to separate the static creation methods from the classes that implement the `match()` function (similar to what is done in the Factory patterns).